gdal/ogr/ogrsf_frmts/cad/gdalcaddataset.cpp

```diff
@@ -0,0 +1,463 @@
+}
+
+int GDALCADDataset::TestCapability( const char * pszCap )
+{
+    if ( EQUAL(pszCap,ODsCCreateLayer) ||
+         EQUAL(pszCap,ODsCDeleteLayer) )
+    {
+         return FALSE;
+    }
+    else if( EQUAL(pszCap,ODsCCurveGeometries) )
+        return TRUE;
+    else if( EQUAL(pszCap,ODsCMeasuredGeometries) )
+        return TRUE;
+    return FALSE;
+}
+
+char** GDALCADDataset::GetFileList()
+{
+    int i, j;
+    char **papszFileList = GDALDataset::GetFileList();
+
+    /* duplicated papszFileList = CSLAddString( papszFileList, osCADFilename );*/
+    const char * pszPRJFilename = GetPrjFilePath();
+    if(NULL != pszPRJFilename)
+        papszFileList = CSLAddString( papszFileList, pszPRJFilename );
+    
+    for(i = 0; i < poCADFile->getLayersCount(); ++i)
+    {
+        CADLayer &oLayer = poCADFile->getLayer( i );
+        for( j = 0; j < oLayer.getImageCount(); ++j )
+        {
+            CADImage* pImage = oLayer.getImage(j);
+            if(pImage)
+            {                
+                CPLString osImgFilename = pImage->getFilePath(); 
+                if ( CPLCheckForFile ((char*)osImgFilename.c_str(), NULL) == TRUE)
+                    papszFileList = CSLAddString( papszFileList, osImgFilename );
+            }
+        }
+    }
+    
+    if(NULL != poRasterDS)
+    {
+        papszFileList = CSLMerge(papszFileList, poRasterDS->GetFileList());
+    }
+    return papszFileList;
+}
+
+OGRSpatialReference *GDALCADDataset::GetSpatialReference()
+{
+    OGRSpatialReference *poSpatialRef = NULL;
+    if( poCADFile != NULL )
+    {
+	CPLString sESRISpatRef;
+        poSpatialRef = new OGRSpatialReference();
+	CADDictionary oNOD = poCADFile->getNOD();
+	for( size_t i = 0; i < oNOD.getRecordsCount(); ++i )
+	{
+		if( !strcmp( oNOD.getRecord(i).first.c_str(), "ESRI_PRJ" ) )
+		{
+			CADXRecord * poXRecord = ( CADXRecord* ) oNOD.getRecord(i).second;
+			// FIXME: optimize it with std::string::find() func
+			size_t dDataBegins = poXRecord->getRecordData().find("GEO");
+
+			std::string sESRISpatRefData( poXRecord->getRecordData().begin() + dDataBegins,
+						      poXRecord->getRecordData().end() );
+			sESRISpatRef = sESRISpatRefData;
+		}
+	}
+
+        if( !sESRISpatRef.empty() )
+        {
+            char** papszPRJData = NULL;
+            papszPRJData = CSLAddString(papszPRJData, sESRISpatRef);
+            if( poSpatialRef->importFromESRI( papszPRJData ) != OGRERR_NONE )
+            {
+                CPLError( CE_Warning, CPLE_AppDefined,
+                        "Failed to parse PRJ section, ignoring." );
+                delete( poSpatialRef );
+                poSpatialRef = NULL;
+            }
+
+            CSLDestroy( papszPRJData );
+        }
+        else
+        {
+            const char * pszPRJFilename = GetPrjFilePath();
+            if(NULL != pszPRJFilename)
+            {
+            CPLPushErrorHandler( CPLQuietErrorHandler );
+            char **cabyPRJData = CSLLoad(pszPRJFilename);
```

Comment:
naming convention: should be papszPRJData (Pointer to an Array of Pointers of Strings Zero-terminated)